### ISSUE TITLE
ZipArchive's __repr__ is wrong in a really confusing way

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3592]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3592|
|Type|enhancement|
|Created|2008-12-24 08:54:10Z|
|Branch|https://github.com/twisted/twisted/tree/zippath-repr-3592|

As shown here:

```
#!py
>>> from twisted.python.zippath import ZipArchive
>>> ZipArchive('x.zip')
ZipPath('x.zip')
```

Attachments:

* [zippath-ZipArchive-repr.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73d/73d5ee840660b355b3cab7e512d16da84cff15fd/5233257a81b7b45513a297d7f6ceb1200c4940aa.patch) (377 bytes) - added by jesstess on 2009-08-18 00:03:47Z - 
* [zippath-reprs.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73d/73d5ee840660b355b3cab7e512d16da84cff15fd/46a3b9f99bdf99ce600f671fe00df0e60ead2208.patch) (670 bytes) - added by jesstess on 2009-08-18 02:54:38Z - modifications to both the ZipArchive and ZipPath reprs
* [test_path_ZipFilePathTestCase.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73d/73d5ee840660b355b3cab7e512d16da84cff15fd/3e6e139778669149cef37678bb576a057983b847.patch) (2113 bytes) - added by jesstess on 2009-08-25 17:52:51Z - 
* [test_paths_worksOnWin.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/73d/73d5ee840660b355b3cab7e512d16da84cff15fd/c17569e3f28985ba112517efd73974dfa0061f36.patch) (3055 bytes) - added by jesstess on 2009-09-22 01:01:57Z - tested on a 64 bit Windows 7 machine

<details><summary>Searchable metadata</summary>

```
trac-id__3592 3592
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_zippath_repr_3592 branches/zippath-repr-3592
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1230108850000000 1230108850000000
changetime__1260753488000000 1260753488000000
version__None None
owner__ 
cc__jesstess cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

What would be right and not confusing?  `ZipArchive('/foo/bar/x.zip')`?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_1):
Yes, that sounds right.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_3"></a>|@jesstess removed owner|
|-|-|

The patch just adds

    def __repr__(self):
        return 'ZipArchive(%r)' % (os.path.abspath(self.path),)

to the class, producing eg 

>>> from twisted.python import zippath
>>> z = zippath.ZipArchive("foo.zip")
>>> z
ZipArchive('/Users/jesstess/foo.zip')

Should ZipPath's repr return an absolute path as well?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @jesstess|
|-|-|

Replying to [jesstess](#note_3):
> The patch just adds
>
>     def __repr__(self):
>         return 'ZipArchive(%r)' % (os.path.abspath(self.path),)
>
> to the class, producing eg
>
> >>> from twisted.python import zippath
> >>> z = zippath.ZipArchive("foo.zip")
> >>> z
> ZipArchive('/Users/jesstess/foo.zip')

Thanks a lot!  It's great to have someone looking at these minor cleanup issues; there are lots of parts of Twisted which need attention.  The patch also needs some tests, though.  Can you add those and resubmit for review?

> Should ZipPath's repr return an absolute path as well?

Yes, that would be good.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess removed owner|
|-|-|

zippath-reprs.patch does the same as the previous patch but also has ZipPath's repr return an absolute path.

The test_path patch adds methods for testing ZipArchive and ZipPath, testing the repr for each that results from a ZipArchive created with either an absolute path or a path without the cwd prepended.

Hacking up self.cmn to get a path without the cwd inside the methods probably isn't ideal, but adding any attributes in setUp to use across methods caused other tests in AbstractFilePathTestCase to fail and I wasn't sure what to do about that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @jesstess|
|-|-|

Thanks for adding these tests.  They look pretty good, with only a few minor things I'd suggest changing:

  1. The only underscore in a test method name should be the one in `test_`.  The rest of the name should be camel case as usual (we use `test_fooBar` instead of `testFooBar` because we use `_` to separate common prefixes where names are automatically discovered based on that prefix).  Can you rename the two new test methods to reflect this?
  1. Operators should have a space on either side - eg `self.cmn+'.zip'` in `test_zippath_repr` should instead be `self.cmn + '.zip'`.  Can you add these spaces?
  1. `cmn` was a bad attribute name to begin with.  Changing it is probably out of scope for this ticket (it would make the patch a lot harder to review, at least), but avoiding copying it in the new tests (ie `rel_cmn`) should be easy.  I think the expansion of `cmn` is supposed to be 'common' (but I'm just guessing)?  That would suggest `relativeCommon` as a good variable name to use instead.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess removed owner|
|-|-|

Updated test_path patch to address points 1-3 by exarkun. Hopefully I'll only make these stylistic mistakes once and then they'll be ingrained. :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Looks great.  Thanks a lot. :)  I'll apply these now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [27243]) Branching to 'zippath-repr-3592'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [27244]) Apply zippath-reprs.patch and test_path_ZipFilePathTestCase.patch

refs [#3592](https://github.com/twisted/twisted/issues/3592)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @jesstess<br>@exarkun set status to `new`|
|-|-|

Oops.  I had a feeling there might be something in this code that would behave differently on Windows, so I tried the tests on Windows before committing to trunk.  Indeed, `twisted.test.test_paths.ZipFilePathTestCase.test_zipPathRepr` fails:

```
===============================================================================
[FAIL]: twisted.test.test_paths.ZipFilePathTestCase.test_zipPathRepr

Traceback (most recent call last):
  File "c:\buildbot\twisted\WXP32-full2.4-scmikes-select\Twisted\twisted\test\test_paths.py", line 298, in test_zipPathRepr
    "ZipPath('%s/%s')" % (os.path.abspath(self.cmn + '.zip'), 'foo'))
twisted.trial.unittest.FailTest: not equal:
a = "ZipPath('c:\\\\buildbot\\\\twisted\\\\WXP32-full2.4-scmikes-select\\\\Twisted\\\\_trial_temp\\\\twisted.test.test_paths\\\\ZipFilePathTestCase\\\\test_zipPathRepr\\\\fx-08-\\\\temp.zip\\\\foo')"
b = "ZipPath('c:\\buildbot\\twisted\\WXP32-full2.4-scmikes-select\\Twisted\\_trial_temp\\twisted.test.test_paths\\ZipFilePathTestCase\\test_zipPathRepr\\fx-08-\\temp.zip/foo')"

-------------------------------------------------------------------------------
```

The two patches are in the branch mentioned above.  If you can make future diffs against that branch instead of trunk, it would be helpful.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_12"></a>|@jesstess removed owner|
|-|-|

os.sep and careful use of reprs has this passing on at least a Windows 7 machine now.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|z3p commented|
|-|-|

(In [27365]) Apply test_paths_worksOnWin.patch

Refs [#3592](https://github.com/twisted/twisted/issues/3592)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|z3p set owner to @jesstess|
|-|-|

Works for me on Windows XP.

I made a couple small tweaks (path -> pathRepr, and a docstring change), and it looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_15"></a>|@jesstess set status to `closed`|
|-|-|

(In [27768]) Merge zippath-repr-3592

Author: jesstess
Reviewers: glyph, exarkun, z3p
Fixes: [#3592](https://github.com/twisted/twisted/issues/3592)

Change the ZipArchive repr to be less confusing, and make the ZipPath
repr return an absolute path for consistency with ZipArchive.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
